README.rst
Spectral Python (SPy)
---------------------
Spectral Python (SPy) is a pure Python module for processing hyperspectral image
data (imaging spectroscopy data). It has functions for reading, displaying,
manipulating, and classifying hyperspectral imagery. Full details about the
package are on the `web site <http://spectralpython.net>`_.


Installation Instructions
=========================

The latest release is always hosted on `PyPI <https://pypi.python.org/pypi/spectral>`_,
so if you have `pip` installed, you can install SPy from the command line with

.. code::

    pip install spectral

Packaged distributions are also hosted at `PyPI`
and `Sourceforge <https://sourceforge.net/projects/spectralpython/files/SPy/>`_
so you can download and unpack the latest zip/tarball, then type

.. code::

    python setup.py install

To install the latest development version, download or clone the git repository
and install as above. No explicit installation is required so you can simply
access (or symlink) the `spectral` module within the source tree.

Unit Tests
==========

To run the suite of unit tests, you must have `numpy` installed and you must
have the `sample data files <http://spectralpython.net/user_guide_intro.html>`_
downloaded to the current directory (or one specified by the `SPECTRAL_DATA`
environment variable). To run the unit tests, type

.. code::

    python -m spectral.tests.run

Dependencies
============
Using SPy interactively with its visualization capabilities requires `IPython` and
several other packages (depending on the features used). See the
`web site <http://spectralpython.net>`_ for details.



CHANGELOG.txt
##############################################################################
# Spectral Python (SPy) package change log
##############################################################################

---------------------------------------------------------------------------
NOTE: This file is now deprecated. Relevant change information can be found
in the git commit logs.
---------------------------------------------------------------------------

2014-03-20	tboggs <tboggs@users.sourceforge.net>

    * io/envi.py (save_image): When saving an image, assume nbands==1 when
      len(img.shape)==2.

2014-03-18	tboggs <tboggs@users.sourceforge.net>

    Added the Minimum Noise Fraction transform.

    * __init__.py: Incremented __version__ to 0.15.0.dev-2
    * algorithms/algorithms.py:
      (noise_from_diffs): New function.
      (MNFResult): New class.
      (mnf): New function.
    * tests/__init__.py: Include MNF tests.
    * tests/dimensionality.py: New file.

2014-03-18	tboggs <tboggs@users.sourceforge.net>

    [closes #2] Handle image data containing NaN. Rather than just passing
    image data through to SPy functions, several checks are now made to
    either warn or raise an exception when `np.nan` is found in the data

    * algorithms/algorithms.py (calc_stats): New argument `allow_nan` (False
      by default) indicates whether nan should be allowed in the data. If
      False, a NanValueError is raised when nan is present.
    * algorithms/clustering.py:
      (kmeans): Make sure status percentage ends in all code paths
      (kmeans_ndarray): Make sure status percentage ends in all code paths.
      Raise NanValueError if nan is present.
    * algorithms/spymath.py:
      (NaNValueWarning): New class
      (NaNValueError): New class
      (has_nan): New function to determine if np.nan is present in the data.
    * io/spyfile.py (SpyFile.load): Issue NaNValueWarning if np.nan is present
      in the image data.

2014-03-11	tboggs <tboggs@users.sourceforge.net>

    [fixed #1] - Pixel row/col is not displayed in imshow window when pixel
    row is greater than image width.

    * graphics/spypylab.py(ImageView.format_coord): `ncols` -> `nrows`.

2014-02-23	tboggs <tboggs@users.sourceforge.net>

    Updated version to 0.14.

2014-02-21	tboggs <tboggs@users.sourceforge.net>

    If available, use Pillow module instead of PIL.

    * graphics/graphics.py (make_pil_image): First try importing `Image` from
      PIL (Pillow) instead of top-level namespace (Python Imaging Library).

2014-02-19	tboggs <tboggs@users.sourceforge.net>

    Use common color scale on all faces of image cube in `view_cube`.

    * graphics/hypercube.py:
      (HypercubeWindow.load_textures): Do not use auto-scaling for cube face
      color scales.

2014-02-19	tboggs <tboggs@users.sourceforge.net>

    Fixed two bugs in `view_cube`: Front and left sides of the cube were
    flipped left-right and cube width/height was scaled incorrectly for
    extreme aspect ratios.

    * graphics/hypercube.py:
      (HypercubeWindow.draw_cube): Corrected aspect ratio.
      (HypercubeWindow.load_textures): Mirrored front and left faces
      left-right.

2014-02-17	tboggs <tboggs@users.sourceforge.net>

    Search `SPECTRAL_DATA` paths for `AsterDatabase` files.

    * database/aster.py (AsterDatabase.__init__).

2014-02-13	tboggs <tboggs@users.sourceforge.net>

    Fix covariance averaging for MahalanobisDistanceClassifier.

    * algorithms/classifiers.py (MahalanobisDistanceClassifier.train):
      Divide by total number of samples when computing weighted covariance.

    Finished Perceptron Classifier improvements.

    * algorithms/classifiers.py (PerceptronClassifier.train): Return
      Perceptron.train result.
    * algorithms/perceptron.py (PerceptronLayer.randomize_weights):
      Randomize weights so initial hyperplanes intersect the unit box [0, 0].
    * test/classifiers.py: Added perceptron unit tests.

2014-02-11	tboggs <tboggs@users.sourceforge.net>

    Fixed PerceptronClassifier implementation.

    * algorithims/__init__.py: import all from classifiers.py.
    * algorithms/classifiers.py: Moved PerceptronClassifier from perceptron.py.
    * algorithms/perceptron.py: Numerous fixes/improvements. This file is now
      independent of the rest of the SPy package (i.e., it is a pure
      multilayer perceptron implementation without the baggage of the rest of
      the SPy package).

2014-01-05	tboggs <tboggs@users.sourceforge.net>

    Enable displaying `imshow` output in an existing matplotlib figure.

    * graphics/spypylab (imshow): Accept optional `fignum` argument.

2014-01-05	tboggs <tboggs@users.sourceforge.net>

    Updated version to 0.13

    Allow modifying `ImageView` classes even when the `classes` arg was not
    provided.

    * graphics/spypylab.py:
      (ImageViewKeyboardHandler.handle_event): Check ImageView.classes instead
        of ImageView.class_axes.
      (ImageView.init_callbacks): Make class update callback handle `classes`
        being None.
      (ImageView.set_display_mode): Call .show_classes if classes are set but
        .class_axes is not (i.e., if classes were set since last update).

    Added Modified SAM (code contributed by Christian Mielke).

    * algorithms/algorithms.py:
      (spectral_angles): Compute angles in 64 bits. Clip dot products to
      prevent bad values in arccos due to roundoff.
      (msam): New function.

    Removed deprecated functions:

    * algorithms/algorithms.py: Removed principalComponents,
      linearDiscriminant, reduceEigenvectors, logDeterminant, ,
      createTrainingClasses, bhattacharyyaDistance, bDistanceTerms,
      transformImage, spectralAngles.

2013-12-27	tboggs <tboggs@users.sourceforge.net>

    Avoid a bug in `wx.NewId` that can cause exceptions when right-clicking in
    an ND window.

    * graphics/ndwindow.py (MouseMenu.__init__): Cache menu item ID's instead of

spectral/io/erdas.py
#########################################################################
#
#   erdas.py - This file is part of the Spectral Python (SPy) package.
#
#   Copyright (C) 2001-2010 Thomas Boggs
#
#   Spectral Python is free software; you can redistribute it and/
#   or modify it under the terms of the GNU General Public License
#   as published by the Free Software Foundation; either version 2
#   of the License, or (at your option) any later version.
#
#   Spectral Python is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this software; if not, write to
#
#               Free Software Foundation, Inc.
#               59 Temple Place, Suite 330
#               Boston, MA 02111-1307
#               USA
#
#########################################################################
#
# Send comments to:
# Thomas Boggs, tboggs@users.sourceforge.net
#

'''
Functions for reading Erdas files.
'''
# Following description accessed on 2011-01-25 at
#    http://www.pcigeomatics.com/cgi-bin/pcihlp/ERDASWR|IMAGE+FORMAT
#
# The ERDAS image file format contains a header record (128 bytes), followed by
# the image data. The image data is arranged in a Band Interleaved by Line
# (BIL) format. Each file is virtually unlimited in size - the file structure
# allows up to 274 billion bytes. The file consists of 512-byte records.
#
#                 ERDAS IMAGE FILE FORMAT
#  +----------------------------------------------------------+
#  |   Record 1 (bytes 1 to 128) Header                       |
#  |   --------------------------------                       |
#  |                                                          |
#  |    Bytes   Type   Contents                               |
#  |                                                          |
#  |     1- 6    ASCII  Descriptor         (HEAD74 or HEADER) |
#  |     7- 8    I*2    Type of data:0=8 bit /1=4 bit/2=16 bit|
#  |     9-10    I*2    Number of Channels                    |
#  |    11-16           Unused                                |
#  |    17-20    I*4    Number of Pixels, if HEAD74           |
#  |            (R*4    Number of Pixels, if HEADER)          |
#  |    21-24    I*4    Number of Lines,  if HEAD74           |
#  |            (R*4    Number of Lines,  if HEADER)          |
#  |    25-28    I*4    X-coordinate of 1st pixel, if HEAD74  |
#  |            (R*4    X-coordinate of 1st pixel, if HEADER) |
#  |    29-32    I*4    Y-coordinate of 1st pixel, if HEAD74  |
#  |            (R*4    Y-coordinate of 1st pixel, if HEADER) |
#  |    33-88           Unused                                |
#  |    89-90    I*2    Integer which indicates Map type      |
#  |    91-92    I*2    Number of classes in the data set     |
#  |    93-106          Unused                                |
#  |   107-108   I*2    Units of area of each pixel           |
#  |                    0=NONE, 1=ACRE, 2=HECTAR, 3=OTHER     |
#  |   109-112   R*4    Number of pixel area units            |
#  |   113-116   R*4    Map X-coordinate of upper left corner |
#  |   117-120   R*4    Map Y-coordinate of upper left corner |
#  |   121-124   R*4    X-pixel size                          |
#  |   125-128   R*4    Y-pixel size                          |
#  |                                                          |
#  |   Data files values begin in bytes 129 and cross over    |
#  |   record boundaries as necessary.                        |
#  |   Data are arranged in following order:                  |
#  |                                                          |
#  |   L - Lines;  C - Channels;  P - Pixels per line;        |
#  |                                                          |
#  |   Pixels 1 through x of line 1, band 1                   |
#  |   Pixels 1 through x of line 1, band n                   |
#  |                                                          |
#  |   Pixels 1 through x of line 2, band 1                   |
#  |   Pixels 1 through x of line 2, band n                   |
#  |                                                          |
#  |   Pixels 1 through x of line y, band 1                   |
#  |   Pixels 1 through x of line y, band n                   |
#  +----------------------------------------------------------+


def open(file):
    '''
    Returns a SpyFile object for an ERDAS/Lan image file.

    Arguments:

        `file` (str):

            Name of the ERDAS/Lan image data file.

    Returns:

        A SpyFile object for the image file.

    Raises:

        IOError
    '''

    import numpy as np
    import spectral
    from bilfile import BilFile
    from spyfile import find_file_path


    # ERDAS 7.5 headers do not specify byte order so we'll guess little endian.
    # If any of the parameters look weird, we'll try again with big endian.

    class Params:
        pass
    p = Params()
    p.byte_order = 0

    file_path = find_file_path(file)

    lh = read_erdas_lan_header(find_file_path(file))
    if lh["nbands"] < 0 or lh["nbands"] > 512 or \
        lh["ncols"] < 0 or lh["ncols"] > 10000 or \
            lh["nrows"] < 0 or lh["nrows"] > 10000:
        p.byte_order = 1
        lh = read_erdas_lan_header(file_path, 1)

    p.filename = file_path
    p.nbands = lh["nbands"]
    p.ncols = lh["ncols"]
    p.nrows = lh["nrows"]
    p.offset = 128
    if lh["packing"] == 2:
        p.dtype = np.dtype('i2').str
    elif lh["packing"] == 0:
        p.dtype = np.dtype('i1').str
    elif lh["packing"] == 1:
        raise Exception(
            '4-bit data type not supported in SPy ERDAS/Lan format handler.')
    else:
        raise Exception('Unexpected data type specified in ERDAS/Lan header.')
    if spectral.byte_order != 0:
        p.dtype = np.dtype(p.dtype).newbyteorder().str

    return BilFile(p, lh)


def read_erdas_lan_header(fileName, byte_order=0):
    '''Read parameters from a lan file header.

    Arguments:

        fileName (str):

            File to open.

        byte_order (int) [default 0]:

            Specifies whether to read as little (0) or big (1) endian.
    '''
    from exceptions import IOError
    from array import array
    import __builtin__
    import spectral

    f = __builtin__.open(fileName, "rb")

    h = {}
    h["format"] = "lan"
    h["fileName"] = fileName
    h["sizeOfHeader"] = 128

    h["type"] = f.read(6)
    if h["type"] not in ('HEAD74', 'HEADER'):
        raise IOError('Does not look like an ERDAS Lan header.')

    # Read all header data into arrays
    word = array('h')
    dword = array('i')
    float = array('f')
    word.fromfile(f, 2)
    f.seek(16)
    if h["type"] == 'HEAD74':
        dword.fromfile(f, 4)
    else:
        float.fromfile(f, 4)
    f.seek(88)
    word.fromfile(f, 2)
    f.seek(106)
    word.fromfile(f, 1)
    float.fromfile(f, 5)

    if byte_order != spectral.byte_order:
        word.byteswap()
        dword.byteswap()
        float.byteswap()

    # Unpack all header data
    h["packing"] = word.pop(0)
    h["nbands"] = word.pop(0)

    if h["type"] == 'HEAD74':
        h["ncols"] = dword.pop(0)
        h["nrows"] = dword.pop(0)
        h["pixel_xcoord"] = dword.pop(0)
        h["pixel_ycoord"] = dword.pop(0)
    else:
        h["ncols"] = int(float.pop(0))
        h["nrows"] = int(float.pop(0))
        h["pixel_xcoord"] = float.pop(0)
        h["pixel_ycoord"] = float.pop(0)

    h["map_type"] = word.pop(0)
    h["nclasses"] = word.pop(0)
    h["area_unit"] = word.pop(0)
    h["ypixel_size"] = float.pop()
    h["xpixel_size"] = float.pop()
    h["map_ycoord"] = float.pop()
    h["map_xcoord"] = float.pop()
    h["narea_units"] = float.pop()

    f.close()

    return h


def readErdasLanHeader(fileName, byte_order=0):
    warn('readErdasLanHeader has been deprecated.  Use read_erdas_lan_header.',
         DeprecationWarning)
    return read_erdas_lan_header(fileName, byte_order)
